File: backend/app/routers/jobs.py
"""The job-contract operator surface: what ran, what failed, and what it cost.

The master plan's target state asks the job layer for exactly that, and before the
contract there was nowhere to ask it. `scheduled_job_runs` covered the three beat
sweeps; the other 39 tasks left no durable record at all, so "did last night's work
happen" was answered by reading worker logs.

AUTH mirrors the 0080 RLS boundary:

* Reads require any provisioned staff (``view_reports``) - which a portal client does
  NOT hold, so clients are 403'd out of the namespace, mirroring backups/reports.
  Postgres backs that up: neither table has a select policy for a client.
* Cancelling a run, replaying a dead letter and resolving one are LEAD actions
  (owner/admin/manager). Replay in particular deliberately re-runs work that has
  already spent money once.

The mutations write on the PRIVILEGED store after the router's permission check -
the same pattern as the audit worker's status writes - because neither table carries
an authenticated write policy by design.
"""

from __future__ import annotations

import asyncio
from typing import Annotated, Any
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.core.auth import CurrentUser, require_perm, require_role
from app.core.pagination import PageDep
from app.db.job_runs_repo import JobRunsRepoDep, job_runs_store
from app.jobs.celery_task import enqueue
from app.jobs.status import JobStatus
from app.schemas.jobs import (
    CancelRequest,
    DeadLetterResponse,
    InFlightResponse,
    JobRunResponse,
    JobSummaryResponse,
    ReplayResponse,
    ResolveRequest,
)
from app.services.activity import record_activity

router = APIRouter(tags=["jobs"])

# All six staff roles hold view_reports; a portal client does NOT.
ViewReports = Annotated[CurrentUser, Depends(require_perm("view_reports"))]
# Stopping and replaying work is a lead decision.
LeadOnly = Annotated[CurrentUser, Depends(require_role("owner", "admin", "manager"))]

_RUN_NOT_FOUND = HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job run not found")
_DL_NOT_FOUND = HTTPException(
    status_code=status.HTTP_404_NOT_FOUND, detail="Dead letter not found"
)


# --------------------------------------------------------------------------- #
# Reads
# --------------------------------------------------------------------------- #
@router.get("/jobs/summary", response_model=JobSummaryResponse)
async def job_summary(
    repo: JobRunsRepoDep,
    _user: ViewReports,
    window_hours: Annotated[int, Query(ge=1, le=720, alias="windowHours")] = 24,
) -> JobSummaryResponse:
    """Runs and spend per terminal state over a window, plus the open DLQ depth.

    ``degraded`` is its own line. It is never added to a success count, which is the
    whole reason the vocabulary distinguishes the two - a board that folds partial
    outcomes into "completed" is the board this contract was built to replace.
    """
    rows = await asyncio.to_thread(repo.counts_by_status, since_hours=window_hours)
    open_dlq = await asyncio.to_thread(repo.count_open_dead_letters)
    return JobSummaryResponse.from_rows(rows, window_hours=window_hours, open_dlq=open_dlq)


@router.get("/jobs/runs", response_model=list[JobRunResponse])
async def list_job_runs(
    repo: JobRunsRepoDep,
    page: PageDep,
    _user: ViewReports,
    run_status: Annotated[JobStatus | None, Query(alias="status")] = None,
    job_name: Annotated[str | None, Query(alias="jobName")] = None,
    client_id: Annotated[UUID | None, Query(alias="clientId")] = None,
    correlation_id: Annotated[UUID | None, Query(alias="correlationId")] = None,
    needs_attention: Annotated[bool, Query(alias="needsAttention")] = False,
) -> list[JobRunResponse]:
    """The runs board, newest first.

    ``needsAttention=true`` is the view that matters day to day: every terminal run
    that was not a clean success - degraded, blocked, failed - in one list.
    ``correlationId`` reassembles a whole fan-out: one nightly sweep and the eighty
    per-client jobs it enqueued share one id.
    """
    rows = await asyncio.to_thread(
        repo.list_runs,
        status=run_status.value if run_status else None,
        job_name=job_name,
        client_id=str(client_id) if client_id else None,
        correlation_id=str(correlation_id) if correlation_id else None,
        needs_attention=needs_attention,
        limit=page.limit,
        offset=page.offset,
    )
    return [JobRunResponse.from_row(r) for r in rows]


@router.get("/jobs/runs/{run_id}", response_model=JobRunResponse)
async def get_job_run(run_id: UUID, repo: JobRunsRepoDep, _user: ViewReports) -> JobRunResponse:
    row = await asyncio.to_thread(repo.get_run, str(run_id))
    if row is None:
        raise _RUN_NOT_FOUND
    return JobRunResponse.from_row(row)


@router.get("/jobs/in-flight", response_model=list[InFlightResponse])
async def in_flight(repo: JobRunsRepoDep, _user: ViewReports) -> list[InFlightResponse]:
    """Running jobs per (client, queue) - exactly what the concurrency cap acts on.

    This is the answer to "why is that client's work not starting": if a client sits
    at its cap on a queue, its next job is deferred rather than started, and after
    ``max_queue_seconds`` it is honestly blocked rather than deferred forever.
    """
    rows = await asyncio.to_thread(repo.in_flight_by_client)
    return [InFlightResponse.from_row(r) for r in rows]


@router.get("/jobs/dead-letters", response_model=list[DeadLetterResponse])
async def list_dead_letters(
    repo: JobRunsRepoDep,
    page: PageDep,
    _user: ViewReports,
    open_only: Annotated[bool, Query(alias="openOnly")] = True,
) -> list[DeadLetterResponse]:
    """Work the platform accepted and did not deliver.

    Open items come OLDEST first - the opposite of every other feed here, and
    deliberately: the longest-unresolved lost job is the most urgent one, not the
    least.
    """
    rows = await asyncio.to_thread(
        repo.list_dead_letters, open_only=open_only, limit=page.limit, offset=page.offset
    )
    return [DeadLetterResponse.from_row(r) for r in rows]


# --------------------------------------------------------------------------- #
# Lead actions
# --------------------------------------------------------------------------- #
@router.post("/jobs/runs/{run_id}/cancel", response_model=JobRunResponse)
async def cancel_job_run(
    run_id: UUID, body: CancelRequest, repo: JobRunsRepoDep, user: LeadOnly
) -> JobRunResponse:
    """Ask a run to stop.

    COOPERATIVE, not a kill: a Celery task cannot be safely terminated part-way
    through writing to a client's website. This sets a flag - a queued run will never
    start, and a running one stops at its next ``ctx.checkpoint()``. A job that does
    not checkpoint cannot be stopped, which is why anything that loops must.

    A run that has already finished returns 409: overwriting a real outcome with
    "cancelled" would be a fiction, and the ledger's job is to not contain those.
    """
    existing = await asyncio.to_thread(repo.get_run, str(run_id))
    if existing is None:
        raise _RUN_NOT_FOUND

    row = await asyncio.to_thread(job_runs_store().request_cancel, str(run_id), requested_by=user.id)
    if row is None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"This run already finished ({existing['status']}) and cannot be cancelled",
        )

    await record_activity(
        user,
        kind="access",
        action="Cancelled a job run",
        target=f"{row['job_name']} ({run_id})",
        meta=body.reason or None,
        entity_type="client" if row.get("client_id") else None,
        entity_id=str(row["client_id"]) if row.get("client_id") else None,
    )
    return JobRunResponse.from_row(row)


@router.post("/jobs/dead-letters/{dead_letter_id}/replay", response_model=ReplayResponse)
async def replay_dead_letter(
    dead_letter_id: UUID, repo: JobRunsRepoDep, user: LeadOnly
) -> ReplayResponse:
    """Re-run a unit of work that was lost, with its original arguments.

    Two things make this safe to expose:

    1. The replay gets its OWN idempotency key (``replay:<dead_letter_id>``), not the
       original's. Reusing the original key would find the old terminal run and skip
       the work silently - the replay would appear to succeed and do nothing.
    2. ``mark_replayed`` only matches a dead letter that is still open, so a double
       click cannot enqueue the work twice. That matters more here than anywhere else
       in the product: replaying deliberately re-runs something that already spent
       money once.

    The run row is created HERE rather than by the worker, so the operator gets a run
    id immediately and the dead letter can point at it. The worker's own claim finds
    this queued row and starts it - the same path a retry takes.
    """
    dead_letter = await asyncio.to_thread(repo.get_dead_letter, str(dead_letter_id))
    if dead_letter is None:
        raise _DL_NOT_FOUND
    if dead_letter.get("resolved_at") or dead_letter.get("replayed_at"):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="This dead letter has already been replayed or resolved",
        )

    store = job_runs_store()
    replay_key = f"replay:{dead_letter_id}"
    payload: dict[str, Any] = dict(dead_letter.get("payload") or {})
    args = list(payload.get("args") or [])
    kwargs = dict(payload.get("kwargs") or {})

    run_row, created = await asyncio.to_thread(
        store.claim,
        job_name=str(dead_letter["job_name"]),
        task=str(dead_letter.get("task") or ""),
        queue=str(dead_letter["queue"]),
        idempotency_key=replay_key,
        correlation_id=str(dead_letter.get("correlation_id") or dead_letter_id),
        parent_run_id=str(dead_letter["run_id"]) if dead_letter.get("run_id") else None,
        celery_task_id="",
        client_id=str(dead_letter["client_id"]) if dead_letter.get("client_id") else None,
        client_name=str(dead_letter.get("client_name") or ""),
        scope_type=str(dead_letter.get("scope_type") or ""),
        scope_id=str(dead_letter["scope_id"]) if dead_letter.get("scope_id") else None,
        # The running code's spec is the truth about the retry budget; `start()`
        # re-stamps this the moment a worker picks the run up.
        max_attempts=1,
    )
    if not created:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="A replay of this dead letter already exists",
        )

    claimed = await asyncio.to_thread(
        store.mark_replayed,
        str(dead_letter_id),
        replayed_run_id=str(run_row["id"]),
        replayed_by=user.id,
    )
    if claimed is None:
        # Someone replayed it between the read and the write. Do not enqueue.
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="This dead letter was replayed by someone else",
        )

    message_id = await asyncio.to_thread(
        enqueue,
        str(dead_letter["task"]),
        *args,
        correlation_id=str(run_row["correlation_id"]),
        idempotency_key=replay_key,
        **kwargs,
    )

    await record_activity(
        user,
        kind="access",
        action="Replayed a dead-lettered job",
        target=f"{dead_letter['job_name']} ({dead_letter_id})",
        meta=f"run {run_row['id']}",
        entity_type="client" if dead_letter.get("client_id") else None,
        entity_id=str(dead_letter["client_id"]) if dead_letter.get("client_id") else None,
    )
    return ReplayResponse(
        dead_letter_id=str(dead_letter_id),
        run_id=str(run_row["id"]),
        message_id=message_id,
        idempotency_key=replay_key,
    )


@router.post("/jobs/dead-letters/{dead_letter_id}/resolve", response_model=DeadLetterResponse)
async def resolve_dead_letter(
    dead_letter_id: UUID, body: ResolveRequest, repo: JobRunsRepoDep, user: LeadOnly
) -> DeadLetterResponse:
    """Close a dead letter with a written decision.

    The resolution text is required by the schema AND by a CHECK constraint. A queue
    closed with no reasons written is a graveyard - the next person cannot tell "we
    fixed the underlying bug" from "we gave up on this one".
    """
    existing = await asyncio.to_thread(repo.get_dead_letter, str(dead_letter_id))
    if existing is None:
        raise _DL_NOT_FOUND

    row = await asyncio.to_thread(
        job_runs_store().resolve_dead_letter,
        str(dead_letter_id),
        resolution=body.resolution,
        resolved_by=user.id,
    )
    if row is None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="This dead letter is already resolved"
        )

    await record_activity(
        user,
        kind="access",
        action="Resolved a dead-lettered job",
        target=f"{row['job_name']} ({dead_letter_id})",
        meta=body.resolution,
    )
    return DeadLetterResponse.from_row(row)

File: frontend/lib/hooks/jobs.ts
"use client";

// ============================================================
// AIOS · job-contract data hooks (the Operations board)
// Backs the Operations board off the FastAPI /jobs endpoints — the one place the
// platform can answer WHAT RAN, WHAT FAILED, AND WHAT IT COST. Response shapes are
// contract-locked to lib/jobs.ts (the backend emits camelCase via
// serialization_alias), so the JSON drops straight into the existing types.
//
// AUTH mirrors routers/jobs.py exactly:
//   • reads          → require_perm("view_reports") — every staff role, no client
//   • cancel/replay/resolve → require_role(owner|admin|manager) — LEAD ONLY
// Gate the act buttons on `isLeadRole(me.data?.role)` from lib/jobs.ts; the server
// remains the boundary.
//
// LIVE SURFACES POLL. The summary, the runs board and in-flight refetch every 10s
// so a running job's row moves on its own — an operations board that only updates
// on F5 is a screenshot. The single-run detail polls only until the run reaches a
// terminal state, then stops (the audits.ts idiom).
// ============================================================

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "@/lib/api";
import type {
  DeadLetter,
  InFlightRow,
  JobRun,
  JobRunFilters,
  JobSummary,
  ReplayResult,
} from "@/lib/jobs";
import { isTerminalStatus } from "@/lib/jobs";

export const JOBS_KEY = ["jobs"] as const;
export const JOB_SUMMARY_KEY = ["jobs", "summary"] as const;
export const JOB_RUNS_KEY = ["jobs", "runs"] as const;
export const IN_FLIGHT_KEY = ["jobs", "in-flight"] as const;
export const DEAD_LETTERS_KEY = ["jobs", "dead-letters"] as const;

/** The live-board poll. Fast enough that a queued→running→completed hop is seen
 *  while the operator is still looking at the row. */
export const JOBS_POLL_MS = 10_000;

// --- reads -------------------------------------------------------------------

/**
 * The operator headline over a window (GET /jobs/summary?windowHours=).
 *
 * `byStatus` keeps `degraded` as its own line — it is never added to a success
 * count, which is the whole reason the vocabulary distinguishes the two. Read
 * `succeededRuns` (completed only) and `needsAttentionRuns` (degraded + blocked +
 * failed) rather than summing statuses by hand.
 */
export function useJobSummary(windowHours = 24) {
  return useQuery({
    queryKey: [...JOB_SUMMARY_KEY, windowHours] as const,
    queryFn: () => api.get<JobSummary>(`/jobs/summary?windowHours=${windowHours}`),
    refetchInterval: JOBS_POLL_MS,
  });
}

/**
 * The runs board, newest first (GET /jobs/runs).
 *
 * `needsAttention: true` is the day-to-day view: every terminal run that was not a
 * clean success, in one list. `correlationId` reassembles a whole fan-out — one
 * nightly sweep and the eighty per-client jobs it enqueued share one id. The server
 * caps `limit` at 200 and defaults to 50.
 */
export function useJobRuns(filters: JobRunFilters = {}) {
  const params = new URLSearchParams();
  if (filters.status) params.set("status", filters.status);
  if (filters.jobName) params.set("jobName", filters.jobName);
  if (filters.clientId) params.set("clientId", filters.clientId);
  if (filters.correlationId) params.set("correlationId", filters.correlationId);
  if (filters.needsAttention) params.set("needsAttention", "true");
  if (filters.limit !== undefined) params.set("limit", String(filters.limit));
  if (filters.offset !== undefined) params.set("offset", String(filters.offset));
  const qs = params.toString();

  return useQuery({
    queryKey: [...JOB_RUNS_KEY, "list", qs] as const,
    queryFn: () => api.get<JobRun[]>(`/jobs/runs${qs ? `?${qs}` : ""}`),
    refetchInterval: JOBS_POLL_MS,
    // Paging/filtering keeps the previous rows on screen instead of flashing an
    // empty board while the new window loads.
    placeholderData: (prev) => prev,
  });
}

/**
 * ONE run, by id (GET /jobs/runs/{run_id}) — the drawer/detail read.
 *
 * Polls at the live cadence while the run can still change, then stops dead: a
 * finished run is immutable, so re-fetching it forever is pure noise.
 */
export function useJobRun(runId: string) {
  return useQuery({
    queryKey: [...JOB_RUNS_KEY, runId] as const,
    queryFn: () => api.get<JobRun>(`/jobs/runs/${runId}`),
    enabled: Boolean(runId),
    refetchInterval: (query) => {
      const run = query.state.data as JobRun | undefined;
      if (!run) return JOBS_POLL_MS;
      return isTerminalStatus(run.status) ? false : JOBS_POLL_MS;
    },
  });
}

/**
 * Running work per (client, queue) — exactly what the per-client concurrency cap
 * acts on (GET /jobs/in-flight). This is the answer to "why is that client's work
 * not starting": a client at its cap has its next job deferred, and after
 * `max_queue_seconds` honestly blocked rather than deferred forever.
 */
export function useInFlight() {
  return useQuery({
    queryKey: IN_FLIGHT_KEY,
    queryFn: () => api.get<InFlightRow[]>("/jobs/in-flight"),
    refetchInterval: JOBS_POLL_MS,
  });
}

/**
 * Work the platform ACCEPTED and did not deliver (GET /jobs/dead-letters?openOnly=).
 *
 * Open items come OLDEST first — the opposite of every other feed here, and
 * deliberately: the longest-unresolved lost job is the most urgent one, not the
 * least. Each row carries the payload that makes it replayable.
 */
export function useDeadLetters(openOnly = true) {
  return useQuery({
    queryKey: [...DEAD_LETTERS_KEY, openOnly] as const,
    queryFn: () => api.get<DeadLetter[]>(`/jobs/dead-letters?openOnly=${openOnly}`),
  });
}

// --- lead-only mutations -----------------------------------------------------

export type CancelRunInput = {
  runId: string;
  /** Recorded on the activity log (who + why), not on the run row. Max 500 chars. */
  reason?: string;
};

/**
 * Ask a run to stop (POST /jobs/runs/{run_id}/cancel). LEAD ONLY.
 *
 * COOPERATIVE, not a kill: a Celery task cannot be safely torn down part-way
 * through writing to a client's website. This sets a flag — a queued run never
 * starts, a running one stops at its next `ctx.checkpoint()`, and a job that never
 * checkpoints cannot be stopped at all. An already-finished run returns 409 rather
 * than overwriting a real outcome with the fiction "cancelled".
 */
export function useCancelRun() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ runId, reason }: CancelRunInput) =>
      api.post<JobRun>(`/jobs/runs/${runId}/cancel`, { reason: reason ?? "" }),
    onSuccess: () => {
      // ["jobs","runs"] prefix-matches both the list keys and every detail key.
      void qc.invalidateQueries({ queryKey: JOB_RUNS_KEY });
      void qc.invalidateQueries({ queryKey: JOB_SUMMARY_KEY });
      void qc.invalidateQueries({ queryKey: IN_FLIGHT_KEY });
    },
  });
}

/**
 * Re-run a lost unit of work with its original arguments
 * (POST /jobs/dead-letters/{id}/replay). LEAD ONLY.
 *
 * Deliberately re-runs something that already spent money once, so the server makes
 * it safe twice over: the replay gets its OWN idempotency key (`replay:<id>`), and
 * `mark_replayed` only matches a still-open dead letter, so a double click cannot
 * enqueue the work twice (the second returns 409). Returns the new run id
 * immediately — the board can follow the replay from the moment it is accepted.
 */
export function useReplayDeadLetter() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (deadLetterId: string) =>
      api.post<ReplayResult>(`/jobs/dead-letters/${deadLetterId}/replay`),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: DEAD_LETTERS_KEY });
      // A replay creates a real queued run: the board, the headline and the
      // concurrency picture all moved.
      void qc.invalidateQueries({ queryKey: JOB_RUNS_KEY });
      void qc.invalidateQueries({ queryKey: JOB_SUMMARY_KEY });
      void qc.invalidateQueries({ queryKey: IN_FLIGHT_KEY });
    },
  });
}

export type ResolveDeadLetterInput = {
  deadLetterId: string;
  /** REQUIRED and non-blank — enforced by the schema AND a DB CHECK. */
  resolution: string;
};

/**
 * Close a dead letter with a written decision
 * (POST /jobs/dead-letters/{id}/resolve). LEAD ONLY.
 *
 * The resolution text is mandatory on purpose: a queue closed with no reasons
 * written is a graveyard — the next person cannot tell "we fixed the underlying
 * bug" from "we gave up on this one". Send a real sentence, not a placeholder.
 */
export function useResolveDeadLetter() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ deadLetterId, resolution }: ResolveDeadLetterInput) =>
      api.post<DeadLetter>(`/jobs/dead-letters/${deadLetterId}/resolve`, { resolution }),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: DEAD_LETTERS_KEY });
      // `openDeadLetters` on the headline just changed.
      void qc.invalidateQueries({ queryKey: JOB_SUMMARY_KEY });
    },
  });
}

export type ReapResult = {
  /** `ok` (nothing to do) or `degraded` (rows were reaped — real work was lost). */
  status: string;
  reaped: number;
  /** Human-readable: "no stale runs", or "reaped run_audit_job x3, run_content_job x1". */
  detail: string;
};

/**
 * Fail every job run whose worker died without writing an outcome
 * (POST /maintenance/reap-stuck-jobs). OWNER ONLY.
 *
 * `JobRunsStore.start` counts `running` rows against the per-client concurrency cap,
 * so a run left `running` by an OOM kill or a host reboot permanently removes a slot
 * from that client — their queue gets quietly narrower and nothing reports it. The
 * reaper is the only thing that clears one.
 *
 * With cron parked (`beat_schedule = {}`), the `reap-stale-job-runs` schedule does not
 * fire, so THIS is the only caller. That is why it needs a button: until this hook
 * existed, unwedging a client's queue meant an owner-token curl.
 *
 * The endpoint runs the sweep INLINE rather than enqueueing it, because dispatching
 * the repair onto the very worker pool that may be wedged is how "I pressed the button
 * and nothing happened" happens. It answers 503 rather than a 200 with a zero count
 * when the ledger is unreachable: "reaped 0" and "could not look" are different
 * answers and only one of them means the queue is healthy.
 */
export function useReapStuckJobs() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: () => api.post<ReapResult>("/maintenance/reap-stuck-jobs"),
    onSuccess: () => {
      // Reaping writes terminal outcomes onto runs the board is showing as running,
      // and frees concurrency slots, so all three of these moved.
      void qc.invalidateQueries({ queryKey: JOB_RUNS_KEY });
      void qc.invalidateQueries({ queryKey: JOB_SUMMARY_KEY });
      void qc.invalidateQueries({ queryKey: IN_FLIGHT_KEY });
    },
  });
}

// --- automations --------------------------------------------------------------
// Scheduled work an admin controls. The fourteen periodic jobs this platform had
// were switched off wholesale on 2026-08-19 because a static beat schedule is read
// at process start: pausing one, re-timing one, or scoping one to some clients each
// needed a deploy. They are rows now (0118), all seeded paused.

export const AUTOMATIONS_KEY = ["automations"] as const;

export type AutomationCapability = {
  kind: string;
  label: string;
  description: string;
  scope: "platform" | "client";
  /** True when a run can spend metered budget. Shown BEFORE anyone enables it. */
  paid: boolean;
  defaultIntervalSeconds: number;
  needs: string[];
};

export type Automation = {
  id: string;
  name: string;
  kind: string;
  kindLabel: string;
  scope: "platform" | "client";
  paid: boolean;
  params: Record<string, unknown>;
  scheduleKind: "interval" | "cron";
  intervalSeconds: number | null;
  cronExpr: string | null;
  /** "every 30 minutes" / "cron: 0 2 * * *" */
  cadence: string;
  enabled: boolean;
  notifyOnFailure: boolean;
  notifyChannels: Record<string, unknown>;
  nextDueAt: string | null;
  lastFiredAt: string | null;
  /** Null before it has ever run - which is not the same as a run that failed. */
  lastRunId: string | null;
  lastStatus: string | null;
  lastFinishedAt: string | null;
  lastDetail: string;
};

export function useAutomations() {
  return useQuery({
    queryKey: AUTOMATIONS_KEY,
    queryFn: () => api.get<Automation[]>("/automations"),
    refetchInterval: JOBS_POLL_MS,
  });
}

export function useAutomationCapabilities() {
  return useQuery({
    queryKey: [...AUTOMATIONS_KEY, "capabilities"] as const,
    queryFn: () => api.get<AutomationCapability[]>("/automations/capabilities"),
    staleTime: Infinity, // a code-level registry; it cannot change under a session
  });
}

export type AutomationInput = {
  name: string;
  kind: string;
  params?: Record<string, unknown>;
  scheduleKind: "interval" | "cron";
  intervalSeconds?: number | null;
  cronExpr?: string | null;
  enabled?: boolean;
  notifyOnFailure?: boolean;
};

export function useCreateAutomation() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (input: AutomationInput) => api.post<Automation>("/automations", input),
    onSuccess: () => void qc.invalidateQueries({ queryKey: AUTOMATIONS_KEY }),
  });
}

/** Edit, pause or resume. Pausing is this call with `enabled: false`. */
export function useUpdateAutomation() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, changes }: { id: string; changes: Partial<AutomationInput> }) =>
      api.patch<Automation>(`/automations/${id}`, changes),
    onSuccess: () => void qc.invalidateQueries({ queryKey: AUTOMATIONS_KEY }),
  });
}

export function useDeleteAutomation() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => api.del<void>(`/automations/${id}`),
    onSuccess: () => void qc.invalidateQueries({ queryKey: AUTOMATIONS_KEY }),
  });
}

/** Fire it once now without changing its schedule - how an automation is tested
 *  before being enabled, and how a missed window is recovered. */
export function useRunAutomationNow() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id: string) =>
      api.post<{ automationId: string; runId: string | null; dispatched: number }>(
        `/automations/${id}/run`,
        {},
      ),
    onSuccess: () => void qc.invalidateQueries({ queryKey: AUTOMATIONS_KEY }),
  });
}
